runners/core-java/src/test/java/org/apache/beam/runners/core/StateInternalsTest.java

```diff
@@ -610,4 +624,40 @@ public void testMapReadable() throws Exception {
     assertThat(value.get("C").read(), equalTo(3));
   }
 
+  @Test
+  public void testBagWithBadCoderEquality() throws Exception {
+    // Ensure two instances of the bad coder are distinct
+    assertThat(new BadStringUtf8Coder(), not(equalTo(new BadStringUtf8Coder())));
+
+    BagState<String> state1 = underTest.state(NAMESPACE_1, STRING_BAG_ADDR1);
+    state1.add("hello");
+
+    BagState<String> state2 = underTest.state(NAMESPACE_1, STRING_BAG_ADDR2);
+    assertThat(state2.read(), containsInAnyOrder("hello"));
+  }
+
+  private static class BadStringUtf8Coder extends Coder<String> {
```

Comment:
`BadEquality...`?

Maybe have an equals and hashCode that throw instead of just assuming they're bad?